### PR TITLE
Ex13 Israel.C DB Jooq exercise

### PR DESCRIPTION
**Pull request by IsraelCohen54**
Ex13 Israel.C DB Jooq exercise

### CODE REVIEW

File: src/main/resources/db/migration/V202409041458__create_table_owners.sql [hunk 1]
```diff
@@ -2,15 +2,15 @@ CREATE TABLE owner (
...
 
-
+/*
 INSERT INTO owner (name, company_id, employee_id)
 VALUES ('Josh', '9', 123123123);
 
 INSERT INTO owner (name, company_id, employee_id)
 VALUES ('Ariel', '9', 999123123);
-
+*/
```
Comment (omermelamed) [inline src/main/resources/db/migration/V202409041458__create_table_owners.sql]: Why is this file relevant for the pr?

File: src/main/resources/db/migration/V202409041445__create_table_pets.sql [hunk 2]
```diff
@@ -29,3 +29,4 @@ WHERE id = 1;
 SELECT *
 FROM pets
 WHERE date_of_arrival < CURRENT_DATE - INTERVAL '1 year';
```
Comment (omermelamed) [inline src/main/resources/db/migration/V202409041445__create_table_pets.sql]: Why is this file relevant for the pr?

File: src/main/kotlin/com/hibob/academy/dao/OwnersDao.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import jakarta.inject.Inject
+
+@Component
+class OwnerDao @Inject constructor(private val sql: DSLContext) {
+
+    private val ownerTableInstance = OwnerTable.ownerInstance
+
+    private val ownerMapper = RecordMapper<Record, Owner> { record ->
+        Owner(
+            //id = record[ownerTableInstance.id],
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/OwnersDao.kt]: But each owner has an Id right? Why did you comment out?

File: src/main/kotlin/com/hibob/academy/dao/OwnersDao.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            .fetchOne(ownerMapper)
+    */
+
+    fun getOwners(companyId: Long): List<Owner?> =
+        sql.selectFrom(ownerTableInstance)
+            .where(ownerTableInstance.companyId.eq(companyId))
+            .fetch(ownerMapper)
+
+    data class Owner(
+        //val id: Long,
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/OwnersDao.kt]: Same here, Each Owner should have an ID

File: src/main/kotlin/com/hibob/academy/dao/OwnersDao.kt [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            //.doNothing()
+            .execute()
+    }
+
+    /*
+    fun readOwner(id: Long): Owner? =
+        sql.selectFrom(ownerTableInstance)
+            .where(ownerTableInstance.id.eq(id))
+            .fetchOne(ownerMapper)
+    */
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/OwnersDao.kt]: If you are not gonna use it or its not relevant for the pr so please remove it.

File: src/main/kotlin/com/hibob/academy/dao/OwnersDao.kt [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        )
+    }
+
+    fun insertOwner(name: String, companyId: Long, employeeId: String) {
+        sql.insertInto(ownerTableInstance)
+            .columns(ownerTableInstance.name, ownerTableInstance.companyId, ownerTableInstance.employeeId)
+            .values(name, companyId, employeeId)
+            //.onConflict(ownerTableInstance.name, ownerTableInstance.companyId, ownerTableInstance.employeeId)
+            //.doNothing()
+            .execute()
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/OwnersDao.kt]: Why did you comment out?
What will happen when you try to insert the same owner twice?

File: src/main/kotlin/com/hibob/academy/dao/OwnersDao.kt [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            //id = record[ownerTableInstance.id],
+            name = record[ownerTableInstance.name],
+            companyId = record[ownerTableInstance.companyId],
+            employeeId = record[ownerTableInstance.employeeId]
+        )
+    }
+
+    fun insertOwner(name: String, companyId: Long, employeeId: String) {
+        sql.insertInto(ownerTableInstance)
+            .columns(ownerTableInstance.name, ownerTableInstance.companyId, ownerTableInstance.employeeId)
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/OwnersDao.kt]: Why do you use .columns and .values and not .set for each column?

File: src/main/kotlin/com/hibob/academy/dao/OwnersDao.kt [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+import org.jooq.DSLContext
+import org.jooq.Record
+import org.jooq.RecordMapper
+import org.springframework.stereotype.Component
+import jakarta.inject.Inject
+
+@Component
+class OwnerDao @Inject constructor(private val sql: DSLContext) {
+
+    private val ownerTableInstance = OwnerTable.ownerInstance
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/OwnersDao.kt]: lets rename it to ownerTable, no need the instance

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+//
+
+@Component
+class PetDao @Inject constructor(private val sql: DSLContext) {
+
+    private val petTableInstance = PetTable.petInstance
+
+    private val petTableMapper = RecordMapper<Record, Pet> { record ->
+        Pet(
+            // id = record[petTableInstance.id],
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: Why comment out? Every Pet should have an Id, Same as the ower.

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+            type = record[petTableInstance.type],
+            companyId = record[petTableInstance.companyId],
+            dateOfArrival = record[petTableInstance.dateOfArrival]//.substring(0,3)
+        )
+    }
+
+
+    fun createPet(name: String, type: String, companyId: Long, dateOfArrival: java.sql.Date) {
+        sql.insertInto(petTableInstance)
+            .columns(petTableInstance.name, petTableInstance.type, petTableInstance.companyId, petTableInstance.dateOfArrival)
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: Same comment, why do you use .columns and .values and not .set for each column?

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+            // id = record[petTableInstance.id],
+            name = record[petTableInstance.name],
+            type = record[petTableInstance.type],
+            companyId = record[petTableInstance.companyId],
+            dateOfArrival = record[petTableInstance.dateOfArrival]//.substring(0,3)
+        )
+    }
+
+
+    fun createPet(name: String, type: String, companyId: Long, dateOfArrival: java.sql.Date) {
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: java.sql.Date?
Lets use LocalDate type instead.

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+            .execute()
+    }
+
+    /*
+    fun getPet(id: Long): Pet? =
+        sql.selectFrom(petTableInstance)
+            .where(petTableInstance.id.eq(id))
+            .fetchOne(petTableMapper) // Use the correct mapper here
+    */
+
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: remove from pr if not relevant.

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+            .fetch(petTableMapper)
+    }
+
+    enum class PetTypes{
+        Dog,
+        Cat
+    }
+
+    data class Pet(
+        // val id: Long,
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: same comment as above.

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+        val name: String,
+        val type: String,
+        val companyId: Long,
+        val dateOfArrival: java.sql.Date
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: java.sql.Date?
Lets use LocalDate instead

File: src/main/kotlin/com/hibob/academy/dao/Tables.kt [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+class OwnerTable(tableName: String = "owner") : JooqTable(tableName) {
+    val id = createBigIntField("id")
+    val name = createVarcharField("name")
+    val companyId = createBigIntField("company_id")
+    val employeeId = createVarcharField("employee_id")
+
+    companion object {
+        val ownerInstance = OwnerTable()
+    }
+}
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/Tables.kt]: Nice!

File: src/main/kotlin/com/hibob/academy/dao/OwnersDao.kt (hunk 4 above)
Comment (omermelamed) [reply src/main/kotlin/com/hibob/academy/dao/OwnersDao.kt]: Please move this date class to `models` file so you wont need to import it in other fields under the same package.

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 12 above)
Comment (omermelamed) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: Please move this date class to `models` file so you wont need to import it in other fields under the same package.

File: src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+// column - owneridtopets
+// employee_is - string
+// owner_id - uuid
+
+@BobDbTest
+class OwnerDaoTest @Inject constructor(private val sql: DSLContext)  {
+
+    private val table = OwnerTable()
+    private val companyId = Random.nextLong()
+    private val oD = OwnerDao(sql)
```
Comment (omermelamed) [inline src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt]: Lets give it a better name.

File: src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+import org.junit.jupiter.api.AfterEach
+import org.junit.jupiter.api.Test
+
+import org.junit.jupiter.api.Assertions.*
+import org.junit.jupiter.api.BeforeEach
+import kotlin.random.Random
+
+// column - owneridtopets
+// employee_is - string
+// owner_id - uuid
```
Comment (omermelamed) [inline src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt]: What is this?

File: src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Test
+    fun `insert owner size test`() {
+        val owner1 = OwnerDao.Owner("Bob1", companyId, "123")
+        val owner2 = OwnerDao.Owner("Bob2", companyId, "124")
+        val owner3 = OwnerDao.Owner("Bob3", companyId, "125")
+        oD.insertOwner(owner1.name, companyId, owner1.employeeId)
+        oD.insertOwner(owner2.name, companyId, owner2.employeeId)
+        oD.insertOwner(owner3.name, companyId, owner3.employeeId)
+        assertEquals(3, oD.getOwners(companyId).size)
```
Comment (omermelamed) [inline src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt]: What do we test here?
Looks like we are testing the same as the first test.

File: src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Test
+    fun `get owner test without any owner`() {
+        val ownerVacantList : List<OwnerDao.Owner?> = emptyList()
+        assertEquals(ownerVacantList, oD.getOwners(companyId))
+    }
+
```
Comment (omermelamed) [inline src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt:77]: I can think of more tests, mainly on the getOwner function.
Lets add more tests.

File: src/test/kotlin/com/hibob/academy/dao/PetDaoTest.kt [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Test
+    fun `get pets by type test without any pets`() {
+        val emptyList: List<PetDao.Pet> = emptyList()
+
+        val fetchedPets = petDao.getPetsByType(PetDao.PetTypes.Dog, companyId)
+
+        assertEquals(emptyList, fetchedPets)
+    }
+
```
Comment (omermelamed) [inline src/test/kotlin/com/hibob/academy/dao/PetDaoTest.kt:61]: Nice!

**Comment by omermelamed** [comment]
@IsraelCohen54 please remove the controller from this pr.
This pr should contain only Dao

File: src/main/resources/db/migration/V202409041458__create_table_owners.sql (hunk 1 above)
Comment (IsraelCohen54) [reply src/main/resources/db/migration/V202409041458__create_table_owners.sql]: it's not... updated the remote

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+        Pet(
+            id = record[petTable.id],
+            name = record[petTable.name],
+            type = record[petTable.type],
+            companyId = record[petTable.companyId],
+            dateOfArrival = record[petTable.dateOfArrival]
+        )
+    }
+
+    fun createPet(name: String, type: String, companyId: Long, dateOfArrival: LocalDate) {
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: Let keep use PetType instead of string and convert it to string in the function.

File: src/main/kotlin/com/hibob/academy/dao/models.kt [hunk 22]
```diff
@@ -1,3 +1,25 @@
...
+
+data class Pet(
+    val id: Long,
+    val name: String,
+    val type: String,
+    val companyId: Long,
+    val dateOfArrival: LocalDate
+)
+
+enum class PetTypes{
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/models.kt]: PetType instead if PetTypes since each pet has 1 type.

File: src/main/resources/db/migration/V202409041445__create_table_pets.sql [hunk 23]
```diff
@@ -8,7 +8,7 @@ CREATE TABLE pets (
 
 CREATE INDEX idx_pets_with_company_id ON pets (company_id);
 
-
+/*
```
Comment (omermelamed) [inline src/main/resources/db/migration/V202409041445__create_table_pets.sql]: Remove from pr please

File: src/main/resources/db/migration/V202409041445__create_table_pets.sql [hunk 24]
```diff
@@ -29,3 +29,4 @@ WHERE id = 1;
 SELECT *
 FROM pets
 WHERE date_of_arrival < CURRENT_DATE - INTERVAL '1 year';
+*/
```
Comment (omermelamed) [inline src/main/resources/db/migration/V202409041445__create_table_pets.sql]: Remove from pr please

File: src/main/resources/db/migration/V202409041458__create_table_owners.sql [hunk 25]
```diff
@@ -1,16 +1,16 @@
...
-INSERT INTO owner (name, company_id, employee_id)
+/*
+INSERT INTO owners (name, company_id, employee_id)
 VALUES ('Josh', '9', 123123123);
 
-INSERT INTO owner (name, company_id, employee_id)
+INSERT INTO owners (name, company_id, employee_id)
 VALUES ('Ariel', '9', 999123123);
-
+*/
```
Comment (omermelamed) [inline src/main/resources/db/migration/V202409041458__create_table_owners.sql]: Should be in a different pr, Yael explained in class how to do it.
Please reach out to her.

File: src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+        val ownerVacantList : List<Owner?> = emptyList()
+        assertEquals(ownerVacantList, ownerDao.getOwners(companyId))
+    }
+
+    @Test
+    fun `insert same owner - do not add to DB using conflict`() {
+        val owner1 = Owner(1, "Bob1", companyId, "123")
+        ownerDao.insertOwner(owner1.name, companyId, owner1.employeeId)
+        ownerDao.insertOwner(owner1.name, companyId, owner1.employeeId)
+        assertEquals(1,ownerDao.getOwners(companyId).size)
```
Comment (omermelamed) [inline src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt]: Maybe the better assert should be like this -
Fetch and check that the owner is the old owner and not the new you tried to add.

File: src/main/resources/db/migration/V202409041458__create_table_owners.sql (hunk 25 above)
Comment (IsraelCohen54) [reply src/main/resources/db/migration/V202409041458__create_table_owners.sql]: ok (I have asked about this before via the slack)

File: src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt (hunk 26 above)
Comment (IsraelCohen54) [reply src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt]: but they'r the same, how can I differentiate...?

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 21 above)
Comment (IsraelCohen54) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: Missed it! thanks

File: src/main/kotlin/com/hibob/academy/dao/models.kt (hunk 22 above)
Comment (omermelamed) [reply src/main/kotlin/com/hibob/academy/dao/models.kt]: Why are you marking resolve if not resolve?

File: src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @Test
+    fun `insert same owner by unique key, changed name value, should not be added to DB using conflict`() {
+        val owner1 = Owner(1, "Bob1", companyId, "123")
+        ownerDao.insertOwner(owner1.name, companyId, owner1.employeeId)
+        ownerDao.insertOwner("other name, same unique key", companyId, owner1.employeeId)
+        assertEquals(1,ownerDao.getOwners(companyId).size)
+        val owners = ownerDao.getOwners(companyId)
+        assertTrue(owners.any { it?.name == owner1.name }, "Owner with name ${owner1.name} does not exist")
```
Comment (omermelamed) [inline src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt]: This test is not really readable, the better test will be to compare the data class that return using assertEqual.

File: src/test/kotlin/com/hibob/academy/dao/PetDaoTest.kt [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+    @Test
+    fun `insert pet test`() {
+
+        petDao.insertPet(name = "Rex", type = PetTypes.Dog.toString(), companyId = companyId, dateOfArrival = LocalDate.of(2024,1, 1))
+        petDao.insertPet(name = "Rexi", type = PetTypes.Dog.toString(), companyId = companyId, dateOfArrival = LocalDate.of(2024,1, 2))
+
+        val fetchedPets = petDao.getPetsByType(PetTypes.Dog, companyId)
+
+        assertEquals(2, fetchedPets.size)
```
Comment (omermelamed) [inline src/test/kotlin/com/hibob/academy/dao/PetDaoTest.kt]: Why not test the data class that returned? the size really tells me anything?

File: src/test/kotlin/com/hibob/academy/dao/PetDaoTest.kt [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+    fun `get pets by type test`() {
+
+        petDao.insertPet(name = "Rex", type = PetTypes.Dog.toString(), companyId = companyId, dateOfArrival = LocalDate.of(2024,1,1))
+        petDao.insertPet(name = "Whiskers", type = PetTypes.Cat.toString(), companyId = companyId, dateOfArrival = LocalDate.of(2024,2,2))
+
+        val fetchedDogs = petDao.getPetsByType(PetTypes.Dog, companyId)
+        val fetchedCats = petDao.getPetsByType(PetTypes.Cat, companyId)
+
+        assertEquals(1, fetchedDogs.size)
+        assertEquals(1, fetchedCats.size)
```
Comment (omermelamed) [inline src/test/kotlin/com/hibob/academy/dao/PetDaoTest.kt]: Why not test the data class that returned? the size really tells me anything?

File: src/main/kotlin/com/hibob/academy/dao/models.kt (hunk 22 above)
Comment (omermelamed) [reply src/main/kotlin/com/hibob/academy/dao/models.kt]: @IsraelCohen54 ?

**Review by omermelamed — APPROVED** [review]

File: src/main/kotlin/com/hibob/academy/dao/models.kt (hunk 22 above)
Comment (IsraelCohen54) [reply src/main/kotlin/com/hibob/academy/dao/models.kt]: I remember changing it, it might not be saved, anyway, changed it...

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 30]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    fun getPetId(name: String, type: PetType, companyId: Long, dateOfArrival: LocalDate) :Long? {
+        return sql.select(petTable.id)
+            .from(petTable)
+            .where(petTable.name.eq(name)
+            .and(petTable.companyId.eq(companyId))
+            .and(petTable.type.eq(type.toString()))
+            .and(petTable.dateOfArrival.eq(dateOfArrival)))
+            .fetchOne(petTable.id)
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: Since we dont have any unique constraints on this table then we can get multiple rows with the same data ( besides the id), meaning you can get exceptions here for too many rows.
Maybe use limit to get the first but im not sure it will return the right id all the time.

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 31]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    fun adopt(petId: Long, ownerId: Long) {
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: CompanyId as well

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+            .where(petTable.id.eq(petId))
+            .execute()
+    }
+
+    fun getOwnerByPet(petId: Long): Long? {
+        return sql.select(petTable.ownerId)
+            .from(petTable)
+            .where(petTable.id.eq(petId))
+            .fetchOne(petTable.ownerId)
+    }
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: You should return Owner and not the id

Comment (IsraelCohen54) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: changed

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 31 above)
Comment (IsraelCohen54) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: added

File: src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt (hunk 27 above)
Comment (IsraelCohen54) [reply src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt]: changed

File: src/test/kotlin/com/hibob/academy/dao/PetDaoTest.kt (hunk 28 above)
Comment (IsraelCohen54) [reply src/test/kotlin/com/hibob/academy/dao/PetDaoTest.kt]: changed

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 30 above)
Comment (IsraelCohen54) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: added - .limit(1),
not sure of a better solution, but a very good point indeed.

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 33]
```diff
@@ -0,0 +1,97 @@
...
+            .where(ownerTable.id.eq(ownerId))
+            .fetchOne { record ->
+                Owner(
+                    id = record[ownerTable.id],
+                    name = record[ownerTable.name],
+                    employeeId = record[ownerTable.employeeId],
+                    companyId = record[ownerTable.companyId]
+                )
+            }
+    }
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: Why not use join operation on the 2 tables and use the owner mapper?

Comment (IsraelCohen54) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: updated

File: src/main/kotlin/com/hibob/academy/dao/OwnersDao.kt [hunk 34]
```diff
@@ -0,0 +1,62 @@
...
+                    .and(ownerTable.id.eq(id)) // Convert `id` to Long if necessary
+            )
+            .fetchOne { record ->
+                Owner(
+                    id = record[ownerTable.id],
+                    name = record[ownerTable.name],
+                    employeeId = record[ownerTable.employeeId],
+                    companyId = record[ownerTable.companyId]
+                )
+            }
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/OwnersDao.kt]: Why not use the mapper?

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 35]
```diff
@@ -0,0 +1,86 @@
...
+        return sql.select(petTable.ownerId)
+            .from(petTable)
+            .where(petTable.id.eq(petId))
+            .fetchOne(petTable.ownerId)
+    }
+
+    fun getPetOwner(petId: Long): Owner? {
+        return sql.select()
+            .from(petTable)
+            .join(ownerTable)
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt:74]: Do we need this join operation or a different one?
What are the differences between all the joins?

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 36]
```diff
@@ -0,0 +1,86 @@
...
+            .on(petTable.ownerId.eq(ownerTable.id))
+            .where(petTable.id.eq(petId))
+            .fetchOne { record ->
+                Owner(
+                    id = record[ownerTable.id],
+                    name = record[ownerTable.name],
+                    employeeId = record[ownerTable.employeeId],
+                    companyId = record[ownerTable.companyId]
+                )
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt:83]: Lets try to avoid duplicating code.

**Review by omermelamed — COMMENTED** [review]
Pay attention you need to add cleanup to one of your dao test, when ever you decide to create the getOwnerByPetId

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 35 above)
Comment (IsraelCohen54) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt:74]: Inner Join: Includes only rows where there is a match in both tables.
Left Join: Includes all rows from the left table (petTable), with matching rows from the right table (ownerTable). Rows from the left table with no match in the right table have NULL values for the right table’s columns.
Right Join: Includes all rows from the right table (ownerTable), with matching rows from the left table (petTable). Rows from the right table with no match in the left table have NULL values for the left table’s columns.

-> as inner row gives 1 relevant row only, and the others give more than once, I think this is the best - I get one row of pet & owner row, and from it, get the owner data. the right and left join would give more rows, which seems unnecessary, isn't it?

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 36 above)
Comment (IsraelCohen54) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt:83]: As said, currently, I have decided to continue with it (open to change...)

Comment (omermelamed) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt:83]: Ok

**Review by omermelamed — APPROVED** [review]
